File: tools/dbench/answers/go_sonnet/s1/r3_report.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

type Order struct {
	ID     int     `json:"id"`
	Amt    float64 `json:"amt"`
	Status string  `json:"status"`
}

func main() {
	cfgData, err := os.ReadFile("app.cfg")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	var cfg map[string]interface{}
	if err := json.Unmarshal(cfgData, &cfg); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	name, _ := cfg["name"].(string)

	ordData, err := os.ReadFile("orders.json")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	var orders []Order
	if err := json.Unmarshal(ordData, &orders); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	paidCount := 0
	totalPaid := 0.0
	for _, o := range orders {
		if o.Status == "paid" {
			paidCount++
			totalPaid += o.Amt
		}
	}
	fmt.Printf("%s: %d paid, total %g\n", name, paidCount, totalPaid)
}

File: tools/dbench/answers/rust_haiku/s3/r3_report.rs
use std::fs;

fn extract_json_string(json: &str, key: &str) -> Option<String> {
    let pat = format!("\"{}\":", key);
    if let Some(pos) = json.find(&pat) {
        let after = &json[pos + pat.len()..].trim_start();
        if after.starts_with('"') {
            let val = &after[1..];
            if let Some(q) = val.find('"') {
                return Some(val[..q].to_string());
            }
        }
    }
    None
}

fn main() {
    let cfg_data = fs::read_to_string("app.cfg").unwrap_or_default();
    let svc_name = extract_json_string(&cfg_data, "name").unwrap_or_default();

    let ord_data = fs::read_to_string("orders.json").unwrap_or_default();
    let mut p_count = 0;
    let mut p_amt = 0.0;

    let buf = ord_data.trim_start_matches('[').trim_end_matches(']');
    for item in buf.split("},{") {
        let obj = item.trim_matches(|c| c == '{' || c == '}');
        let mut amt = 0.0;
        let mut st = "";

        for kv in obj.split(',') {
            let kv = kv.trim();
            if kv.starts_with("\"amt\"") {
                if let Some(i) = kv.find(':') {
                    amt = kv[i + 1..].trim().parse::<f64>().unwrap_or(0.0);
                }
            } else if kv.starts_with("\"status\"") {
                if let Some(i) = kv.find(':') {
                    st = kv[i + 1..].trim().trim_matches('"');
                }
            }
        }

        if st == "paid" {
            p_count += 1;
            p_amt += amt;
        }
    }

    println!("{}: {} paid, total {}", svc_name, p_count, p_amt);
}
